### PR TITLE
add template search module and update inference module with it

### PR DESCRIPTION
**Pull request by NikiforovSergei**

### CODE REVIEW

File: kb/link_test.scs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+@contour0 = [*
+	@link0 = [Драсте];;
```
Comment (KovalM) [inline kb/link_test.scs:2]: Use English

File: problem-solver/cxx/CMakeLists.txt [hunk 2]
```diff
@@ -1 +1,4 @@
+set(OSTIS_INFERENCE_CXX_SRC "${CMAKE_CURRENT_LIST_DIR}")
+
 add_subdirectory(inferenceModule)
+add_subdirectory(templateSearchModule)
```
Comment (KovalM) [inline problem-solver/cxx/CMakeLists.txt:4]: Let's discuss about templateSearchModule. I propose not creating separate module for it

File: problem-solver/cxx/inferenceModule/manager/DirectInferenceManager.cpp [hunk 3]
```diff
@@ -85,41 +92,38 @@ bool DirectInferenceManager::useRule(ScAddr const & rule, vector<ScAddr> const &
   SC_LOG_DEBUG("Created " + to_string(ifStatementParamsList.size()) + " statement params variants");
   for (const auto& ifStatementParams : ifStatementParamsList)
   {
-    ScTemplate ifStatementTemplate;
-    if (ms_context->HelperBuildTemplate(ifStatementTemplate, ifStatement, ifStatementParams))
+    templateSearchModule::TemplateSearcher templateSearchManager(ms_context);
```
Comment (KovalM) [inline problem-solver/cxx/inferenceModule/manager/DirectInferenceManager.cpp:95]:
```suggestion
    templateSearchModule::TemplateSearcher templateSearcher(ms_context);
```

Comment (KovalM) [reply problem-solver/cxx/inferenceModule/manager/DirectInferenceManager.cpp:95]: Can we create TemplateSearcher in the constructor?

File: problem-solver/cxx/templateSearchModule/searcher/TemplateSearcher.cpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  vector<ScTemplateSearchResultItem> searchResult;
+  ScTemplate searchTemplate;
+  if (context->HelperBuildTemplate(searchTemplate, templateAddr, templateParams))
+  {
+    if (context->HelperCheckEdge(
+          TemplateSearchKeynodes::concept_template_with_links,
+          templateAddr,
+          ScType::EdgeAccessConstPosPerm))
+    {
+      searchResult = searchTemplateWithContent(searchTemplate, templateAddr);
```
Comment (KovalM) [inline problem-solver/cxx/templateSearchModule/searcher/TemplateSearcher.cpp:37]: Why we can't pass template params?

File: problem-solver/cxx/templateSearchModule/searcher/TemplateSearcher.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    {
+      std::string stringContent;
+      ScStreamConverter::StreamToString(context->GetLinkContent(link), stringContent);
+      linksContent.emplace(context->HelperGetSystemIdtf(link), stringContent);
+    }
+  }
+  return linksContent;
+}
+
+ScTemplateSearchResultItem TemplateSearcher::getSearchResultItem(const ScTemplateSearchResult & searchResult, const int & itemIndex)
```
Comment (KovalM) [inline problem-solver/cxx/templateSearchModule/searcher/TemplateSearcher.cpp:107]: Why we need create new SearchResultItem?

**Review by KovalM — CHANGES_REQUESTED** [review]

File: problem-solver/cxx/templateSearchModule/searcher/TemplateSearcher.cpp (hunk 5 above)
Comment (NikiforovSergei) [reply problem-solver/cxx/templateSearchModule/searcher/TemplateSearcher.cpp:107]: this is an unused method. forgot to delete

File: problem-solver/cxx/inferenceModule/InferenceModule.hpp [hunk 6]
```diff
@@ -9,11 +9,11 @@
 #include <sc-memory/sc_memory.hpp>
 #include <sc-memory/sc_module.hpp>
 
-#include "inferenceModule.generated.hpp"
+#include "InferenceModule.generated.hpp"
 
 class InferenceModule : public ScModule
 {
-  SC_CLASS(LoadOrder(50))
+  SC_CLASS(LoadOrder(100))
```
Comment (KovalM) [inline problem-solver/cxx/inferenceModule/InferenceModule.hpp:16]: Why we need 100?

**Review by KovalM — APPROVED** [review]

**Pull request merged**